File: 2020/day04.py
import re

data = { 'byr': lambda x: '1920' <= x <= '2002',
         'iyr': lambda x: '2010' <= x <= '2020',
         'eyr': lambda x: '2020' <= x <= '2030',
         'hgt': lambda x: '59' <= x[:-2] <= '76' if 'in' in x else '150' <= x[:-2] <= '193' if x[-2:] in ['cm','in'] else False,
         'hcl': lambda x: re.search(r'^#(?:[0-9a-fA-F]{3}){1,2}$', x),
         'ecl': lambda x: x in ['amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'],
         'pid': lambda x: len(x)==9 and not re.search(r'\D', x) }

def validate_passport_fields(passports):
    print(len([0 for p in passports if all(f in p.keys() for f in data.keys())]))

def validate_passport_data(passports):
    print(len([0 for p in passports if all(f in p.keys() for f in data.keys()) and all(data[k](v) for k, v in p.items() if k!='cid')]))

if __name__ == "__main__":
    with open('2020/input/day04.txt') as f:

        passports = [{(kv:=pp.split(':'))[0]:kv[1] for pp in p.replace('\n', ' ').split(' ')} for p in f.read().split('\n\n')]
                
        validate_passport_fields(passports) # 239
        validate_passport_data(passports)   # 188

File: 2020/day04.js
require('fs').readFile("./2020/input/day04.txt", "utf8", (err, data) => {
    console.log('part 1:' + [...data.replace(/\r/g, '').matchAll(/(^((byr:[^\s]+)|(iyr:[^\s]+)|(eyr:[^\s]+)|(hgt:[^\s]+)|(hcl:[^\s]+)|(ecl:[^\s]+)|(pid:[^\s]+))( |\n(?!\n))((byr:[^\s]+)|(iyr:[^\s]+)|(eyr:[^\s]+)|(hgt:[^\s]+)|(hcl:[^\s]+)|(ecl:[^\s]+)|(pid:[^\s]+)|( |\n(?!\n))){11})|(^((byr:[^\s]+)|(iyr:[^\s]+)|(eyr:[^\s]+)|(hgt:[^\s]+)|(hcl:[^\s]+)|(ecl:[^\s]+)|(pid:[^\s]+)|(cid:[^\s]+))( |\n(?!\n))((byr:[^\s]+)|(iyr:[^\s]+)|(eyr:[^\s]+)|(hgt:[^\s]+)|(hcl:[^\s]+)|(ecl:[^\s]+)|(pid:[^\s]+)|(cid:[^\s]+)|( |\n(?!\n))){13})/gm)].length)
    console.log('part 2:' + [...data.replace(/\r/g, '').matchAll(/(^((byr:(19[2-8][0-9]|199[0-9]|200[0-2]))|(iyr:(201[0-9]|2020))|(eyr:(202[0-9]|2030))|(hgt:((1[5-8][0-9]|19[0-3])cm|(59|6[0-9]|7[0-6])in))|(hcl:#[0-9a-f]{6})|(ecl:(amb|blu|brn|gry|grn|hzl|oth))|(pid:\d{9}))( |\n(?!\n))((byr:(19[2-8][0-9]|199[0-9]|200[0-2]))|(iyr:(201[0-9]|2020))|(eyr:(202[0-9]|2030))|(hgt:((1[5-8][0-9]|19[0-3])cm|(59|6[0-9]|7[0-6])in))|(hcl:#[0-9a-f]{6})|(ecl:(amb|blu|brn|gry|grn|hzl|oth))|(pid:\d{9})|( |\n(?!\n))){11})|(^((byr:(19[2-8][0-9]|199[0-9]|200[0-2]))|(iyr:(201[0-9]|2020))|(eyr:(202[0-9]|2030))|(hgt:((1[5-8][0-9]|19[0-3])cm|(59|6[0-9]|7[0-6])in))|(hcl:#[0-9a-f]{6})|(ecl:(amb|blu|brn|gry|grn|hzl|oth))|(pid:\d{9})|(cid:[\w\d]+))( |\n(?!\n))((byr:(19[2-8][0-9]|199[0-9]|200[0-2]))|(iyr:(201[0-9]|2020))|(eyr:(202[0-9]|2030))|(hgt:((1[5-8][0-9]|19[0-3])cm|(59|6[0-9]|7[0-6])in))|(hcl:#[0-9a-f]{6})|(ecl:(amb|blu|brn|gry|grn|hzl|oth))|(pid:\d{9})|(cid:[\w\d]+)|( |\n(?!\n))){13})/gm)].length)
})